src/framework/constraints.ml

```diff
@@ -507,16 +507,54 @@ struct
 end
 
 module NoContext = struct let name = "no context" end
-module IntConf =
-struct
-  let n () = max_int
-  let names x = Format.asprintf "%d" x
+
+module type GasVal = sig
+  val n: unit -> int
+end
+
+module type Gas = sig
+  module M:Lattice.S
+  val startgas: unit -> M.t
+  val is_exhausted: fundec -> M.t -> bool
+  val is_any_exhausted: M.t -> bool
+  val callee_gas: fundec -> M.t -> M.t
+  val thread_gas: varinfo -> M.t -> M.t
 end
 
+module GlobalGas(GasVal:GasVal):Gas = struct
+  module M = Lattice.Chain (struct include GasVal let names x = Format.asprintf "%d" x end)
+  let startgas () = M.top () (*  get_int "ana.context.gas_value" *)
+
+  let is_any_exhausted v = v <= 0
+  let is_exhausted _  = is_any_exhausted
+
+  (* callee gas = caller gas - 1 *)
+  let callee_gas f v = max 0 (v - 1)
+  let thread_gas f v = max 0 (v - 1)
+end
+
+module PerFunctionGas(GasVal:GasVal):Gas = struct
+  module G = Lattice.Chain (struct include GasVal let names x = Format.asprintf "%d" x end)
+  module M = MapDomain.MapTop_LiftBot(CilType.Fundec)(G)
+  let startgas () = M.empty ()
+  let is_exhausted f v = GobOption.exists (fun g -> g <= 0) (M.find_opt f v)  (* v <= 0 *)
+  let is_any_exhausted v = M.exists (fun _ g -> g <=0) v
```

Comment:
For programs with many functions, this may become a bit expensive if called often. How often will this be called, given that it is used in the query?

Maybe adding an entry to the data structure that maintains the information whether any of the entries is `0` could be usedul?

Reply:
I left a longer comment below, but basically it is called once per function call, provided the gas is active, and 
threadflag is normally context-sensitive unless gas is exhausted.

The data structure to memoize this would need to be quite complicated as the join may replenish the gas for any of these functions, so I am not sure if that is worth it.

However, I just realized it may be possible to localize this by giving he `JoinCall` another argument.

Reply:
I have removed this now.